pkg/canary/deployment_controller.go

```diff
@@ -364,18 +380,40 @@ func (c *DeploymentController) reconcilePrimaryHpa(cd *flaggerv1.Canary, init bo
 	if !init && primaryHpa != nil {
 		diffMetrics := cmp.Diff(hpaSpec.Metrics, primaryHpa.Spec.Metrics)
 		diffBehavior := cmp.Diff(hpaSpec.Behavior, primaryHpa.Spec.Behavior)
-		if diffMetrics != "" || diffBehavior != "" || int32Default(hpaSpec.MinReplicas) != int32Default(primaryHpa.Spec.MinReplicas) || hpaSpec.MaxReplicas != primaryHpa.Spec.MaxReplicas {
+		diffLabels := cmp.Diff(hpa.ObjectMeta.Labels, primaryHpa.ObjectMeta.Labels)
+		diffAnnotations := cmp.Diff(hpa.ObjectMeta.Annotations, primaryHpa.ObjectMeta.Annotations)
+		if diffMetrics != "" || diffBehavior != "" || diffLabels != "" || diffAnnotations != "" || int32Default(hpaSpec.MinReplicas) != int32Default(primaryHpa.Spec.MinReplicas) || hpaSpec.MaxReplicas != primaryHpa.Spec.MaxReplicas {
 			fmt.Println(diffMetrics, diffBehavior, hpaSpec.MinReplicas, primaryHpa.Spec.MinReplicas, hpaSpec.MaxReplicas, primaryHpa.Spec.MaxReplicas)
-			hpaClone := primaryHpa.DeepCopy()
-			hpaClone.Spec.MaxReplicas = hpaSpec.MaxReplicas
-			hpaClone.Spec.MinReplicas = hpaSpec.MinReplicas
-			hpaClone.Spec.Metrics = hpaSpec.Metrics
-			hpaClone.Spec.Behavior = hpaSpec.Behavior
 
-			_, err := c.kubeClient.AutoscalingV2beta2().HorizontalPodAutoscalers(cd.Namespace).Update(context.TODO(), hpaClone, metav1.UpdateOptions{})
+			err = retry.RetryOnConflict(retry.DefaultRetry, func() error {
+				primaryHpa, err := c.kubeClient.AutoscalingV2beta2().HorizontalPodAutoscalers(cd.Namespace).Get(context.TODO(), primaryHpaName, metav1.GetOptions{})
+				if err != nil {
+					return err
+				}
+				hpaClone := primaryHpa.DeepCopy()
+				hpaClone.Spec.MaxReplicas = hpaSpec.MaxReplicas
+				hpaClone.Spec.MinReplicas = hpaSpec.MinReplicas
+				hpaClone.Spec.Metrics = hpaSpec.Metrics
+				hpaClone.Spec.Behavior = hpaSpec.Behavior
+
+				// update hpa annotations
+				hpaClone.ObjectMeta.Annotations = make(map[string]string)
+				for k, v := range hpa.ObjectMeta.Annotations {
+					hpaClone.ObjectMeta.Annotations[k] = v
```

Comment:
Wouldn't this break HPA custom metrics? As far as I know the autoscaler tracks metrics by saving the last read in the an annotation, if we copy it from the canary, then the primary pods will scaled down.

Reply:
I updated it to reuse `includeLabelsByPrefix` to also filter annotations. This seems preferable to an "ignore" list because it's easier to maintain. Thoughts?

Were you thinking of: https://pkg.go.dev/k8s.io/kubernetes/pkg/apis/autoscaling?utm_source=gopls#DropRoundTripHorizontalPodAutoscalerAnnotations ?